File: web/src/app/models/movie.ts
import {LibraryFile} from '@app/models/library-file';

export interface Genre {
  id: number;
  name: string;
}

export interface Cast {
  cast_id: number;
  character: string;
  credit_id: string;
  gender?: number;
  id: number;
  name: string;
  order: number;
  profile_path?: string;
}

export interface Crew {
  credit_id: string;
  department: string;
  gender?: number;
  id: number;
  job: string;
  name: string;
  profile_path?: string;
}

export interface MovieDetails {
  id: number;
  budget: number;
  genres: Genre[];
  revenue: number;
  runtime?: number;
  tagline?: string;
  credits: {
    cast: Cast[];
    crew: Crew[];
  };
}

export interface Movie {
  id: number;
  title: string;
  original_title: string;
  original_language: string;
  release_date: string;
  poster?: string;
  backdrop?: string;
  overview: string;
  vote_average: number;
  files: LibraryFile[];
  details?: MovieDetails;
}

File: src/main/scala/net/easyflix/app/Movie.scala
package net.easyflix.app

import net.easyflix.json.JsonSupport
import net.easyflix.tmdb.common.{Credits, Genre}
import spray.json.RootJsonFormat

case class Movie(
    id: Int,
    title: String,
    original_title: String,
    original_language: String,
    release_date: String,
    poster: Option[String],
    backdrop: Option[String],
    overview: String,
    vote_average: Float,
    files: Set[LibraryFile],
    details: Option[Movie.Details]) {

  def withDetails(details: Movie.Details): Movie = copy(details = Some(details))

  def withFiles(files: Set[LibraryFile]): Movie =
    copy(files = this.files ++ files.filter(f => !this.files.map(_.path).contains(f.path)))

}

object Movie extends JsonSupport {

  implicit val format: RootJsonFormat[Movie] = jsonFormat11(Movie.apply)

  case class Details(
      id: Int,
      budget: Int,
      genres: List[Genre],
      revenue: Int,
      runtime: Option[Int],
      tagline: Option[String],
      credits: Credits)

  object Details {

    implicit val format: RootJsonFormat[Details] = jsonFormat7(Details.apply)

  }

}
